File: src/software/ai/hl/stp/tactic/goalie/goalie_tactic_test.cpp
#include "software/ai/hl/stp/tactic/goalie/goalie_tactic.h"

#include <gtest/gtest.h>

#include <utility>

#include "software/ai/hl/stp/tactic/move/move_tactic.h"
#include "software/simulated_tests/non_terminating_validation_functions/enemy_never_scores_validation.h"
#include "software/simulated_tests/simulated_er_force_sim_tactic_test_fixture.h"
#include "software/simulated_tests/terminating_validation_functions/ball_kicked_validation.h"
#include "software/simulated_tests/terminating_validation_functions/robot_in_polygon_validation.h"
#include "software/simulated_tests/terminating_validation_functions/robot_received_ball_validation.h"
#include "software/simulated_tests/validation/validation_function.h"
#include "software/test_util/test_util.h"
#include "software/time/duration.h"
#include "software/world/field.h"
#include "software/world/world.h"

class GoalieTacticTest
    : public SimulatedErForceSimTacticTestFixture,
      public ::testing::WithParamInterface<std::tuple<BallState, RobotStateWithId>>
{
   protected:
    void checkGoalieSuccess(int seconds_to_wait, std::shared_ptr<World> world_ptr,
                            ValidationCoroutine::push_type& yield)
    {
        auto initial_time = world_ptr->getMostRecentTimestamp();
        while (world_ptr->getMostRecentTimestamp() <
               initial_time + Duration::fromSeconds(seconds_to_wait))
        {
            yield("Waiting " + std::to_string(seconds_to_wait) +
                  " seconds to check that the enemy team did not score");
        }
        while (contains(world_ptr->field().friendlyDefenseArea(),
                        world_ptr->ball().position()))
        {
            yield("Ball is in the friendly defense area");
        }
    }
    TbotsProto::FieldType field_type = TbotsProto::FieldType::DIV_B;
    Field field                      = Field::createField(field_type);
    std::vector<RobotStateWithId> enemy_robots =
        TestUtil::createStationaryRobotStatesWithId(
            {Point(1, 0), Point(1, 2.5), Point(1, -2.5), field.enemyGoalCenter(),
             field.enemyDefenseArea().negXNegYCorner(),
             field.enemyDefenseArea().negXPosYCorner()});
    std::shared_ptr<const AiConfig> ai_config =
        std::make_shared<ThunderbotsConfig>()->getAiConfig();
};

TEST_F(GoalieTacticTest, test_panic_ball_very_fast_in_straight_line)
{
    BallState ball_state(Point(0, 0), Vector(-3, 0));
    auto friendly_robots = TestUtil::createStationaryRobotStatesWithId({Point(-4, -1)});

    auto tactic = std::make_shared<GoalieTactic>(ai_config);
    setTactic(tactic);
    setFriendlyRobotId(0);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [this, tactic](std::shared_ptr<World> world_ptr,
                       ValidationCoroutine::push_type& yield) {
            checkGoalieSuccess(1, world_ptr, yield);
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            enemyNeverScores(world_ptr, yield);
        }};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(GoalieTacticTest, test_panic_ball_very_fast_in_diagonal_line)
{
    BallState ball_state(Point(0, 0), Vector(-4.5, 0.25));
    auto friendly_robots = TestUtil::createStationaryRobotStatesWithId(
        {field.friendlyGoalCenter() + Vector(0, -0.5)});

    auto tactic = std::make_shared<GoalieTactic>(ai_config);
    setTactic(tactic);
    setFriendlyRobotId(0);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [this, tactic](std::shared_ptr<World> world_ptr,
                       ValidationCoroutine::push_type& yield) {
            checkGoalieSuccess(2, world_ptr, yield);
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            enemyNeverScores(world_ptr, yield);
        }};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(GoalieTacticTest, test_ball_very_fast_misses_net)
{
    BallState ball_state(Point(0, 0), Vector(-4, 1));
    auto friendly_robots = TestUtil::createStationaryRobotStatesWithId({Point(-4.5, 0)});

    auto tactic = std::make_shared<GoalieTactic>(ai_config);
    setTactic(tactic);
    setFriendlyRobotId(0);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [this](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            checkGoalieSuccess(5, world_ptr, yield);
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            enemyNeverScores(world_ptr, yield);
        }};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_F(GoalieTacticTest, test_slow_ball_at_sharp_angle_to_friendly_goal)
{
    BallState ball_state(Point(-4.5, -3), Vector(0, 0.1));
    auto friendly_robots = TestUtil::createStationaryRobotStatesWithId({Point(-4.5, 0)});

    auto tactic = std::make_shared<GoalieTactic>(ai_config);
    setTactic(tactic);
    setFriendlyRobotId(0);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [this](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            checkGoalieSuccess(5, world_ptr, yield);
            robotInPolygon(Rectangle(world_ptr->field().friendlyNegativeYQuadrant()), 1,
                           world_ptr, yield);
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            enemyNeverScores(world_ptr, yield);
        }};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

TEST_P(GoalieTacticTest, goalie_test)
{
    BallState ball_state         = std::get<0>(GetParam());
    RobotStateWithId robot_state = std::get<1>(GetParam());

    auto friendly_robots = {robot_state};

    auto tactic = std::make_shared<GoalieTactic>(ai_config);
    setTactic(tactic);
    setFriendlyRobotId(0);

    std::vector<ValidationFunction> terminating_validation_functions = {
        [this, tactic](std::shared_ptr<World> world_ptr,
                       ValidationCoroutine::push_type& yield) {
            robotReceivedBall(world_ptr, yield);
            checkGoalieSuccess(1, world_ptr, yield);
        }};

    std::vector<ValidationFunction> non_terminating_validation_functions = {
        [](std::shared_ptr<World> world_ptr, ValidationCoroutine::push_type& yield) {
            enemyNeverScores(world_ptr, yield);
        }};

    runTest(field_type, ball_state, friendly_robots, enemy_robots,
            terminating_validation_functions, non_terminating_validation_functions,
            Duration::fromSeconds(10));
}

INSTANTIATE_TEST_CASE_P(
    GoalieEnvironment, GoalieTacticTest,
    ::testing::Values(
        // ball slow inside friendly defense area
        std::make_tuple(BallState(Point(-4, 0.8), Vector(-0.2, 0)),
                        RobotStateWithId{
                            0, RobotState(Point(-4, 0), Vector(0, 0),
                                          Angle::fromDegrees(0), Angle::fromDegrees(0))}),

        // ball stationary inside friendly defense area
        std::make_tuple(
            BallState(Point(-4, 0), Vector(0, 0)),
            RobotStateWithId{
                0,
                RobotState(Field::createSSLDivisionBField().friendlyGoalpostPos(),
                           Vector(0, 0), Angle::fromDegrees(0), Angle::fromDegrees(0))}),

        // ball stationary inside no-chip rectangle
        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter() +
                                      Vector(0.1, 0.1),
                                  Vector(-0.2, 0)),
                        RobotStateWithId{
                            0, RobotState(Point(-4, -1), Vector(0, 0),
                                          Angle::fromDegrees(0), Angle::fromDegrees(0))}),

        // ball fast inside no-chip rectangle but no intersection with goal
        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter() +
                                      Vector(0.1, 0),
                                  Vector(0, -0.5)),
                        RobotStateWithId{
                            0, RobotState(Point(-3.5, 1), Vector(0, 0),
                                          Angle::fromDegrees(0), Angle::fromDegrees(0))}),

        // TODO(#2471): re-enable and fix this
        // ball moving out from inside defense area
        //        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter()
        //        +
        //                                      Vector(0.5, 0),
        //                                  Vector(0.5, 0)),
        //                        RobotStateWithId{
        //                            0, RobotState(Point(-3.5, 0), Vector(0, 0),
        //                                          Angle::fromDegrees(0),
        //                                          Angle::fromDegrees(0))}),

        // TODO(#2471): re-enable and fix this
        // ball moving into goal from inside defense area
        //        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter()
        //        +
        //                                      Vector(0.5, 0),
        //                                  Vector(-0.5, 0)),
        //                        RobotStateWithId{
        //                            0, RobotState(Point(-3.5, 0), Vector(0, 0),
        //                                          Angle::fromDegrees(0),
        //                                          Angle::fromDegrees(0))}),

        // TODO(#2471): re-enable and fix this
        // ball moving up and out of defense area
        //        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter()
        //        +
        //                                      Vector(0.3, 0),
        //                                  Vector(0, 1)),
        //                        RobotStateWithId{
        //                            0, RobotState(Point(-3.5, 0), Vector(0, 0),
        //                                          Angle::fromDegrees(0),
        //                                          Angle::fromDegrees(0))}),

        // ball moving down and out goal from defense area
        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter() +
                                      Vector(0.3, 0),
                                  Vector(0, -0.7)),
                        RobotStateWithId{
                            0, RobotState(Point(-3.5, 0), Vector(0, 0),
                                          Angle::fromDegrees(0), Angle::fromDegrees(0))}),

        // ball slow inside no-chip rectangle
        std::make_tuple(BallState(Field::createSSLDivisionBField().friendlyGoalCenter() +
                                      Vector(0.1, 0),
                                  Vector(0.1, -0.1)),
                        RobotStateWithId{0, RobotState(Point(-3.5, 1), Vector(0, 0),
                                                       Angle::fromDegrees(0),
                                                       Angle::fromDegrees(0))})));

File: src/software/ai/hl/stp/tactic/goalie/goalie_tactic_test.py
import pytest

import software.python_bindings as tbots_cpp
from software.simulated_tests.robot_enters_region import *
from software.simulated_tests.ball_enters_region import *
from software.simulated_tests.ball_moves_forward import *
from software.simulated_tests.friendly_has_ball_possession import *
from software.simulated_tests.ball_speed_threshold import *
from software.simulated_tests.robot_speed_threshold import *
from software.simulated_tests.excessive_dribbling import *
from software.simulated_tests.simulated_test_fixture import (
    simulated_test_runner,
    pytest_main,
)
from proto.message_translation.tbots_protobuf import create_world_state
from proto.ssl_gc_common_pb2 import Team


@pytest.mark.parametrize(
    "ball_initial_position,ball_initial_velocity,robot_initial_position",
    [
        # test panic ball very fast in straight line
        (tbots_cpp.Point(0, 0), tbots_cpp.Vector(-5, 0), tbots_cpp.Point(-4, 0)),
        # test panic ball very_fast in diagonal line
        # TODO (#2609): failing tests when thunderscope is off
        # (
        #     tbots_cpp.Point(0, 0),
        #     tbots_cpp.Vector(-5.5, 0.25),
        #     tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
        #     + tbots_cpp.Vector(0, -0.5),
        # ),
        # test ball very fast misses net
        (tbots_cpp.Point(0, 0), tbots_cpp.Vector(-5, 1), tbots_cpp.Point(-4.5, 0)),
        # test slow ball at sharp angle to friendly goal
        # TODO (#2609): failing tests when thunderscope is off
        # ball slow inside friendly defense area
        # (tbots_cpp.Point(-4, 0.8), tbots_cpp.Vector(-0.2, 0), tbots_cpp.Point(0, 0)),
        # # ball slow inside friendly defense area
        # (tbots_cpp.Point(-4, 0.8), tbots_cpp.Vector(-0.2, 0), tbots_cpp.Point(0, 2)),
        # # ball slow inside friendly defense area
        # (tbots_cpp.Point(-4, 0.8), tbots_cpp.Vector(-0.2, 0), tbots_cpp.Point(0, 2)),
        # ball slow inside friendly defense area
        (tbots_cpp.Point(-4, 0.8), tbots_cpp.Vector(-0.2, 0), tbots_cpp.Point(-4, 0),),
        # ball stationary inside friendly defense area
        (
            tbots_cpp.Point(-4, 0.0),
            tbots_cpp.Vector(0.0, 0),
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalpostPos(),
        ),
        # ball stationary inside no-chip rectangle
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.1, 0.1),
            tbots_cpp.Vector(-0.2, 0),
            tbots_cpp.Point(-4, -1),
        ),
        # ball fast inside no-chip rectangle but no intersection with goal
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.1, 0),
            tbots_cpp.Vector(0, -0.5),
            tbots_cpp.Point(-3.5, 1),
        ),
        # ball moving out from inside defense area
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.5, 0),
            tbots_cpp.Vector(0.5, 0),
            tbots_cpp.Point(-3.5, 0),
        ),
        # ball slow inside no-chip rectangle
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.1, 0),
            tbots_cpp.Vector(0.1, -0.1),
            tbots_cpp.Point(-3.5, 1),
        ),
        # ball moving into goal from inside defense area
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.5, 0),
            tbots_cpp.Vector(-0.5, 0),
            tbots_cpp.Point(-3.5, 0),
        ),
        # ball moving up and out of defense area
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.3, 0),
            tbots_cpp.Vector(0, 1),
            tbots_cpp.Point(-3.5, 0),
        ),
        # ball moving down and out goal from defense area
        (
            tbots_cpp.Field.createSSLDivisionBField().friendlyGoalCenter()
            + tbots_cpp.Vector(0.3, 0),
            tbots_cpp.Vector(0, -0.7),
            tbots_cpp.Point(-3.5, 0),
        ),
    ],
)
def test_goalie_blocks_shot(
    ball_initial_position,
    ball_initial_velocity,
    robot_initial_position,
    simulated_test_runner,
):

    # Setup Robot
    simulated_test_runner.simulator_proto_unix_io.send_proto(
        WorldState,
        create_world_state(
            [],
            blue_robot_locations=[robot_initial_position],
            ball_location=ball_initial_position,
            ball_velocity=ball_initial_velocity,
        ),
    )

    # These aren't necessary for this test, but this is just an example
    # of how to send commands to the simulator.
    #
    # NOTE: The gamecontroller responses are automatically handled by
    # the gamecontroller context manager class
    simulated_test_runner.gamecontroller.send_gc_command(
        gc_command=Command.Type.STOP, team=Team.UNKNOWN
    )
    simulated_test_runner.gamecontroller.send_gc_command(
        gc_command=Command.Type.FORCE_START, team=Team.BLUE
    )

    # Setup Tactic
    params = AssignedTacticPlayControlParams()
    params.assigned_tactics[0].goalie.CopyFrom(
        GoalieTactic(max_allowed_speed_mode=MaxAllowedSpeedMode.PHYSICAL_LIMIT)
    )
    simulated_test_runner.blue_full_system_proto_unix_io.send_proto(
        AssignedTacticPlayControlParams, params
    )

    # Setup no tactics on the enemy side
    params = AssignedTacticPlayControlParams()
    simulated_test_runner.yellow_full_system_proto_unix_io.send_proto(
        AssignedTacticPlayControlParams, params
    )

    # Always Validation
    always_validation_sequence_set = [
        [
            RobotNeverEntersRegion(
                regions=[tbots_cpp.Field.createSSLDivisionBField().enemyDefenseArea()]
            ),
            BallNeverEntersRegion(
                regions=[tbots_cpp.Field.createSSLDivisionBField().friendlyGoal()]
            ),
            NeverExcessivelyDribbles(),
        ]
    ]

    # Eventually Validation
    eventually_validation_sequence_set = [
        [
            # Goalie should be in the defense area
            RobotEventuallyEntersRegion(
                regions=[
                    tbots_cpp.Field.createSSLDivisionBField().friendlyDefenseArea()
                ]
            ),
        ]
    ]

    simulated_test_runner.run_test(
        inv_eventually_validation_sequence_set=eventually_validation_sequence_set,
        inv_always_validation_sequence_set=always_validation_sequence_set,
        ag_eventually_validation_sequence_set=eventually_validation_sequence_set,
        ag_always_validation_sequence_set=always_validation_sequence_set,
    )


if __name__ == "__main__":
    pytest_main(__file__)
